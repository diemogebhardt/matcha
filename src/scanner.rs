use unicode_segmentation::{GraphemeIndices, UnicodeSegmentation};

pub type Range = std::ops::Range<usize>;
pub type Position = usize;

type Iter<'a> = std::iter::Peekable<GraphemeIndices<'a>>;

#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Text(String),
    OpenLine,
    CloseLine,
    OpenValue,
    CloseValue,
    OpenBuilder,
    CloseBuilder,
    IdentifierOrGleamToken(String),
    Import,
    ImportDetails(String),
    With,
    As,
    OpenStmt,
    CloseStmt,
    If,
    Else,
    EndIf,
    For,
    EndFor,
    In,
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let str = match self {
            Token::Text(string) => string,
            Token::OpenLine => "{>",
            Token::CloseLine => "\n",
            Token::OpenValue => "{{",
            Token::CloseValue => "}}",
            Token::OpenBuilder => "{[",
            Token::CloseBuilder => "]}",
            Token::OpenStmt => "{%",
            Token::CloseStmt => "%}",
            Token::IdentifierOrGleamToken(name) => name,
            Token::Import => "import",
            Token::ImportDetails(_) => "import-details",
            Token::With => "with",
            Token::As => "as",
            Token::If => "if",
            Token::Else => "else",
            Token::EndIf => "endif",
            Token::For => "for",
            Token::EndFor => "endfor",
            Token::In => "in",
        };
        write!(f, "{}", str)
    }
}

#[derive(Debug)]
pub enum ScanError {
    UnexpectedGrapheme(String, Position),
    UnexpectedEnd,
}

type Tokens = Vec<(Token, Range)>;

pub fn scan(contents: &str) -> Result<Tokens, ScanError> {
    log::trace!("scan");
    let iter = contents.grapheme_indices(true);

    let tokens = scan_plain(&mut iter.peekable(), vec![])?;

    Ok(tokens)
}

fn scan_plain(iter: &mut Iter, mut tokens: Tokens) -> Result<Tokens, ScanError> {
    log::trace!("scan_plain");
    let mut buffer = String::new();
    let mut buffer_start_index = None;
    let mut buffer_end_index = 0;
    loop {
        match iter.next() {
            Some((first_index, "{")) => {
                if let Some((second_index, "{")) = iter.peek() {
                    if !buffer.is_empty() {
                        tokens.push((
                            Token::Text(buffer),
                            Range {
                                start: buffer_start_index.unwrap_or(0),
                                end: first_index,
                            },
                        ));
                        buffer = String::new();
                    }

                    tokens.push((
                        Token::OpenValue,
                        Range {
                            start: first_index,
                            end: *second_index,
                        },
                    ));
                    iter.next();

                    tokens = scan_identifiers(iter, tokens, |first, second| {
                        matches!(first, Some((_, "}"))) && matches!(second, Some((_, "}")))
                    })?;
                } else if let Some((second_index, "%")) = iter.peek() {
                    if !buffer.is_empty() {
                        tokens.push((
                            Token::Text(buffer),
                            Range {
                                start: buffer_start_index.unwrap_or(0),
                                end: first_index,
                            },
                        ));
                        buffer = String::new();
                    }

                    tokens.push((
                        Token::OpenStmt,
                        Range {
                            start: first_index,
                            end: *second_index,
                        },
                    ));
                    iter.next();

                    tokens = scan_identifiers(iter, tokens, |first, second| {
                        matches!(first, Some((_, "%"))) && matches!(second, Some((_, "}")))
                    })?;
                } else if let Some((second_index, "[")) = iter.peek() {
                    if !buffer.is_empty() {
                        tokens.push((
                            Token::Text(buffer),
                            Range {
                                start: buffer_start_index.unwrap_or(0),
                                end: first_index,
                            },
                        ));
                        buffer = String::new();
                    }

                    tokens.push((
                        Token::OpenBuilder,
                        Range {
                            start: first_index,
                            end: *second_index,
                        },
                    ));

                    iter.next();

                    tokens = scan_identifiers(iter, tokens, |first, second| {
                        matches!(first, Some((_, "]"))) && matches!(second, Some((_, "}")))
                    })?;
                } else if let Some((second_index, ">")) = iter.peek() {
                    tokens.push((
                        Token::OpenLine,
                        Range {
                            start: first_index,
                            end: *second_index,
                        },
                    ));
                    iter.next();

                    tokens = scan_line(iter, tokens)?;

                    let range = consume_grapheme(iter, "\n")?;
                    tokens.push((Token::CloseLine, range));
                } else {
                    buffer.push('{');
                }
            }
            Some((first_index, "}")) => {
                if let Some((second_index, "}")) = iter.peek() {
                    tokens.push((
                        Token::CloseValue,
                        Range {
                            start: first_index,
                            end: *second_index,
                        },
                    ));
                    buffer_start_index = None;
                    iter.next();
                } else {
                    buffer.push('}');
                }
            }
            Some((first_index, "%")) => {
                if let Some((second_index, "}")) = iter.peek() {
                    tokens.push((
                        Token::CloseStmt,
                        Range {
                            start: first_index,
                            end: *second_index,
                        },
                    ));
                    buffer_start_index = None;
                    iter.next();
                } else {
                    buffer.push('%');
                }
            }
            Some((first_index, "]")) => {
                if let Some((second_index, "}")) = iter.peek() {
                    tokens.push((
                        Token::CloseBuilder,
                        Range {
                            start: first_index,
                            end: *second_index,
                        },
                    ));
                    buffer_start_index = None;
                    iter.next();
                } else {
                    buffer.push(']');
                }
            }
            Some((index, grapheme)) => {
                buffer.push_str(grapheme);
                buffer_start_index = buffer_start_index.or(Some(index));
                buffer_end_index = index;
            }
            None => {
                if !buffer.is_empty() {
                    tokens.push((
                        Token::Text(buffer),
                        Range {
                            start: buffer_start_index.unwrap_or(0),
                            end: buffer_end_index,
                        },
                    ));
                }
                return Ok(tokens);
            }
        }
    }
}

type BreakCondition = fn(Option<&(usize, &str)>, Option<&(usize, &str)>) -> bool;

fn scan_identifiers(
    iter: &mut Iter,
    mut tokens: Tokens,
    break_condition: BreakCondition,
) -> Result<Tokens, ScanError> {
    log::trace!("scan_identifiers");

    loop {
        // Clone iterator and move it forward so we can match on
        // the next two graphemes at the same time
        let mut next_grapheme_iter = iter.clone();
        next_grapheme_iter.next();

        match (iter.peek(), next_grapheme_iter.peek()) {
            (first, second) if break_condition(first, second) => break,
            (None, _) => break,
            (Some((_, _)), _) => {
                eat_spaces(iter);
                let (identifier_or_keyword, range) =
                    scan_identifier_or_keyword(iter, break_condition)?;
                tokens.push((identifier_or_keyword, range));
                eat_spaces(iter);
            }
        }
    }

    Ok(tokens)
}

fn scan_identifier_or_keyword(
    iter: &mut Iter,
    break_condition: BreakCondition,
) -> Result<(Token, Range), ScanError> {
    log::trace!("scan_identifier_or_keyword");
    let mut name = String::new();
    let mut start = None;

    // Clone iterator and move it forward so we can match on
    // the next two graphemes at the same time
    let mut next_grapheme_iter = iter.clone();
    next_grapheme_iter.next();

    loop {
        match (iter.peek(), next_grapheme_iter.peek()) {
            (first, second) if break_condition(first, second) => break,
            (Some((_, " ")), _) => break,
            (None, _) => break,
            (Some((index, grapheme)), _) => {
                start = start.or(Some(*index));
                name.push_str(grapheme);
                iter.next();
                next_grapheme_iter.next();
            }
        }
    }

    let trimmed = name.trim();
    let start = start.unwrap_or(0);

    Ok((
        to_token(trimmed),
        Range {
            start,
            end: start + trimmed.len(),
        },
    ))
}

fn to_token(identifier: &str) -> Token {
    log::trace!("to_token: {}", identifier);
    match identifier {
        "if" => Token::If,
        "else" => Token::Else,
        "endif" => Token::EndIf,
        "for" => Token::For,
        "endfor" => Token::EndFor,
        "in" => Token::In,
        "import" => Token::Import,
        "with" => Token::With,
        "as" => Token::As,
        other => Token::IdentifierOrGleamToken(other.to_string()),
    }
}

fn scan_line(iter: &mut Iter, mut tokens: Tokens) -> Result<Tokens, ScanError> {
    eat_spaces(iter);

    let (token, range) =
        scan_identifier_or_keyword(iter, |first, _| matches!(first, Some((_, "\n"))))?;
    tokens.push((token.clone(), range));

    eat_spaces(iter);

    match token {
        Token::Import => {
            let (import_details, range) = scan_import_details(iter);
            tokens.push((Token::ImportDetails(import_details), range));
        }
        _ => {
            tokens = scan_identifiers(iter, tokens, |first, _| matches!(first, Some((_, "\n"))))?;
        }
    }

    Ok(tokens)
}

fn scan_import_details(iter: &mut Iter) -> (String, Range) {
    log::trace!("scan_import_details");
    let mut details = String::new();
    let mut start = None;
    let mut end = 0;

    loop {
        match iter.peek() {
            Some((_index, "\n")) => {
                break;
            }
            Some((index, grapheme)) => {
                start = start.or(Some(*index));
                end = *index + 1;

                details.push_str(grapheme);
                iter.next();
            }
            None => {
                break;
            }
        }
    }

    (
        details,
        Range {
            start: start.unwrap_or(0),
            end,
        },
    )
}

fn consume_grapheme(iter: &mut Iter, expected: &str) -> Result<Range, ScanError> {
    log::trace!("consume_grapheme");
    match iter.next() {
        Some((index, grapheme)) if grapheme == expected => Ok(Range {
            start: index,
            end: index,
        }),
        entry => Err(entry
            .map(|(index, value)| ScanError::UnexpectedGrapheme(value.to_string(), index))
            .unwrap_or(ScanError::UnexpectedEnd)),
    }
}

fn eat_spaces(iter: &mut Iter) {
    log::trace!("eat_spaces");
    while let Some((_, " ")) = iter.peek() {
        iter.next();
    }
}

#[cfg(test)]
mod test {
    use std::fmt::Debug;

    use super::*;

    fn format_debug_result<T: Debug, E: Debug>(result: Result<T, E>) -> String {
        match result {
            Ok(value) => format!("{:#?}", value),
            Err(err) => format!("{:#?}", err),
        }
    }

    #[macro_export]
    macro_rules! assert_scan {
        ($text:expr $(,)?) => {{
            let _ = env_logger::try_init();
            insta::assert_snapshot!(
                insta::internals::AutoName,
                format_debug_result(scan($text)),
                $text
            );
        }};
    }

    // Scan

    #[test]
    fn test_scan_pure_text() {
        assert_scan!("Hello name, good to meet you");
    }

    #[test]
    fn test_scan_identifier() {
        assert_scan!("Hello {{ name }}, good to meet you");
    }

    #[test]
    fn test_scan_two_identifiers() {
        assert_scan!("Hello {{ name }}, {{ adjective }} to meet you");
    }

    #[test]
    fn test_scan_gleam_expression() {
        assert_scan!("Hello {{ string.uppercase(name) }}, good to meet you");
    }

    #[test]
    fn test_scan_single_parens() {
        assert_scan!("Hello { name }, good to meet you");
    }

    #[test]
    fn test_scan_if_statement() {
        assert_scan!("Hello {% if is_user %}User{% endif %}");
    }

    #[test]
    fn test_scan_if_else_statement() {
        assert_scan!("Hello {% if is_user %}User{% else %}Unknown{% endif %}");
    }

    #[test]
    fn test_scan_if_comparison() {
        assert_scan!("Hello {% if items != [] %}Some items{% endif %}");
    }

    #[test]
    fn test_scan_nested_if_statements() {
        assert_scan!(
            "Hello {% if is_user %}{% if is_admin %}Admin{% else %}User{% endif %}{% endif %}"
        );
    }

    #[test]
    fn test_scan_for_loop() {
        assert_scan!("Hello {% for item in list %}{{ item }}{% endfor %}");
    }

    #[test]
    fn test_scan_for_as_loop() {
        assert_scan!("Hello {% for item as Item in list %}{{ item }}{% endfor %}");
    }

    #[test]
    fn test_scan_for_from_expression() {
        assert_scan!("Hello {% for item as Item in list.take(list, 2) %}{{ item }}{% endfor %}");
    }

    #[test]
    fn test_scan_dot_access() {
        assert_scan!("Hello{% if user.is_admin %} Admin{% endif %}");
    }

    #[test]
    fn test_scan_import() {
        assert_scan!("{> import user.{User}\n{{ name }}");
    }

    #[test]
    fn test_scan_with() {
        assert_scan!("{> with user as User\n{{ user }}");
    }

    #[test]
    fn test_scan_builder_block() {
        assert_scan!("Hello {[ builder ]}, good to meet you");
    }
}
