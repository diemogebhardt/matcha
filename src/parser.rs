use crate::scanner::{Range, Token};

type TokenIter<'a> = std::iter::Peekable<std::slice::Iter<'a, (Token, Range)>>;

type Type = String;

#[derive(Debug)]
pub enum Node {
    Text(String),
    Identifier(String),
    Builder(String),
    If(String, Vec<Node>, Vec<Node>),
    For(String, Option<Type>, String, Vec<Node>),
    Import(String),
    With((String, Range), Type),
}

#[derive(Debug)]
pub enum ParserError {
    UnexpectedToken(Token, Range, Vec<Token>),
    UnexpectedEnd,
}

pub fn parse(tokens: &mut TokenIter) -> Result<Vec<Node>, ParserError> {
    log::trace!("parse");
    parse_inner(tokens, false)
}

fn parse_statement(tokens: &mut TokenIter) -> Result<Node, ParserError> {
    match tokens.next() {
        Some((Token::If, _)) => parse_if_statement(tokens),
        Some((Token::For, _)) => parse_for_statement(tokens),
        Some((token, range)) => Err(ParserError::UnexpectedToken(
            token.clone(),
            range.clone(),
            vec![Token::If, Token::For],
        )),
        None => Err(ParserError::UnexpectedEnd),
    }
}

fn parse_inner(tokens: &mut TokenIter, in_statement: bool) -> Result<Vec<Node>, ParserError> {
    log::trace!("parse_inner");
    let mut ast = vec![];

    loop {
        match tokens.next() {
            Some((Token::Text(text), _)) => {
                ast.push(Node::Text(text.clone()));
            }
            Some((Token::OpenValue, _)) => {
                let (name, _) = extract_code(tokens)?;
                ast.push(Node::Identifier(name.clone()));
                consume_token(tokens, Token::CloseValue)?;
            }
            Some((Token::OpenBuilder, _)) => {
                let (name, _) = extract_code(tokens)?;
                ast.push(Node::Builder(name.clone()));
                consume_token(tokens, Token::CloseBuilder)?;
            }
            Some((Token::OpenStmt, _)) => {
                if let Some((Token::Else, _)) | Some((Token::EndIf, _)) | Some((Token::EndFor, _)) =
                    tokens.peek()
                {
                    if in_statement {
                        break;
                    } else {
                        match tokens.next() {
                            Some((token, range)) => {
                                return Err(ParserError::UnexpectedToken(
                                    token.clone(),
                                    range.clone(),
                                    vec![Token::If, Token::For],
                                ))
                            }
                            None => return Err(ParserError::UnexpectedEnd),
                        }
                    }
                }
                let node = parse_statement(tokens)?;
                ast.push(node);
            }
            Some((Token::OpenLine, _)) => {
                match tokens.next() {
                    Some((Token::Import, _)) => {
                        let import_details = extract_import_details(tokens)?;
                        ast.push(Node::Import(import_details))
                    }
                    Some((Token::With, _)) => {
                        let (identifier, range) = extract_identifier(tokens)?;
                        consume_token(tokens, Token::As)?;
                        let (type_, _) = extract_identifier(tokens)?;
                        ast.push(Node::With((identifier, range), type_))
                    }
                    _ => {}
                }
                consume_token(tokens, Token::CloseLine)?;
            }
            Some((token, range)) => {
                return Err(ParserError::UnexpectedToken(
                    token.clone(),
                    range.clone(),
                    vec![],
                ))
            }
            None => {
                break;
            }
        }
    }

    Ok(ast)
}

fn parse_if_statement(tokens: &mut TokenIter) -> Result<Node, ParserError> {
    log::trace!("parse_if_statement");
    let (name, _) = extract_code(tokens)?;
    consume_token(tokens, Token::CloseStmt)?;

    let if_nodes = parse_inner(tokens, true)?;
    let mut else_nodes = vec![];

    match tokens.next() {
        Some((Token::EndIf, _)) => {
            consume_token(tokens, Token::CloseStmt)?;
        }
        Some((Token::Else, _)) => {
            consume_token(tokens, Token::CloseStmt)?;

            else_nodes = parse_inner(tokens, true)?;
            consume_token(tokens, Token::EndIf)?;
            consume_token(tokens, Token::CloseStmt)?;
        }
        Some((token, range)) => {
            return Err(ParserError::UnexpectedToken(
                token.clone(),
                range.clone(),
                vec![Token::EndIf, Token::Else],
            ));
        }
        None => {
            return Err(ParserError::UnexpectedEnd);
        }
    }

    Ok(Node::If(name, if_nodes, else_nodes))
}

fn parse_for_statement(tokens: &mut TokenIter) -> Result<Node, ParserError> {
    let (entry_identifier, _) = extract_identifier(tokens)?;
    let entry_type = match tokens.next() {
        Some((Token::As, _)) => {
            let (type_identifier, _) = extract_identifier(tokens)?;
            consume_token(tokens, Token::In)?;
            Some(type_identifier)
        }
        Some((Token::In, _)) => None,
        Some((matched_token, range)) => {
            return Err(ParserError::UnexpectedToken(
                matched_token.clone(),
                range.clone(),
                vec![Token::As, Token::In],
            ));
        }
        None => return Err(ParserError::UnexpectedEnd),
    };

    let (list_identifier, _) = extract_code(tokens)?;
    consume_token(tokens, Token::CloseStmt)?;

    let loop_nodes = parse_inner(tokens, true)?;

    consume_token(tokens, Token::EndFor)?;
    consume_token(tokens, Token::CloseStmt)?;

    Ok(Node::For(
        entry_identifier,
        entry_type,
        list_identifier,
        loop_nodes,
    ))
}

fn extract_identifier(tokens: &mut TokenIter) -> Result<(String, Range), ParserError> {
    log::trace!("extract_identifier");
    match tokens.next() {
        Some((Token::IdentifierOrGleamToken(name), range)) => Ok((name.clone(), range.clone())),
        Some((token, range)) => Err(ParserError::UnexpectedToken(
            token.clone(),
            range.clone(),
            vec![Token::IdentifierOrGleamToken("".to_string())],
        )),
        None => Err(ParserError::UnexpectedEnd),
    }
}

fn extract_code(tokens: &mut TokenIter) -> Result<(String, Range), ParserError> {
    log::trace!("extract_code");
    let mut code = String::new();
    let mut range: Option<Range> = None;

    loop {
        match tokens.peek() {
            Some((Token::IdentifierOrGleamToken(name), token_range)) => {
                // Create range and expand it to include all the tokens
                // that we're adding to this string
                range = range
                    .map(|current| Range {
                        start: std::cmp::min(token_range.start, current.start),
                        end: std::cmp::max(token_range.end, current.end),
                    })
                    .and(Some(token_range.clone()));

                // Place a space between each identifier
                if !code.is_empty() {
                    code.push(' ');
                }

                code.push_str(name);
                tokens.next();
            }
            Some((Token::CloseStmt, _)) => break,
            Some((Token::CloseValue, _)) => break,
            Some((Token::CloseBuilder, _)) => break,
            Some((token, range)) => {
                if code.is_empty() {
                    return Err(ParserError::UnexpectedToken(
                        token.clone(),
                        range.clone(),
                        vec![Token::IdentifierOrGleamToken("".to_string())],
                    ));
                } else {
                    break;
                }
            }
            None => return Err(ParserError::UnexpectedEnd),
        }
    }

    Ok((code, range.unwrap_or(Range { start: 0, end: 0 })))
}

fn extract_import_details(tokens: &mut TokenIter) -> Result<String, ParserError> {
    log::trace!("extract_import_details");
    match tokens.next() {
        Some((Token::ImportDetails(details), _)) => Ok(details.clone()),
        Some((token, range)) => Err(ParserError::UnexpectedToken(
            token.clone(),
            range.clone(),
            vec![Token::ImportDetails("".to_string())],
        )),
        None => Err(ParserError::UnexpectedEnd),
    }
}

fn consume_token(tokens: &mut TokenIter, expected_token: Token) -> Result<(), ParserError> {
    log::trace!("consume_token");
    match tokens.next() {
        Some((matched_token, _)) if *matched_token == expected_token => Ok(()),
        Some((matched_token, range)) => Err(ParserError::UnexpectedToken(
            matched_token.clone(),
            range.clone(),
            vec![expected_token],
        )),
        None => Err(ParserError::UnexpectedEnd),
    }
}

#[cfg(test)]
mod test {
    use std::fmt::Debug;

    use crate::scanner::{self, ScanError};

    use super::*;

    #[derive(Debug)]
    pub enum Error {
        Scan(ScanError),
        Parse(ParserError),
    }

    fn format_debug_result<T: Debug, E: Debug>(result: Result<T, E>) -> String {
        match result {
            Ok(value) => format!("{:#?}", value),
            Err(err) => format!("{:#?}", err),
        }
    }

    #[macro_export]
    macro_rules! assert_parse {
        ($text:expr $(,)?) => {{
            let _ = env_logger::try_init();
            let result = scanner::scan($text)
                .map_err(|err| Error::Scan(err))
                .and_then(|tokens| {
                    parse(&mut tokens.iter().peekable()).map_err(|err| Error::Parse(err))
                });
            insta::assert_snapshot!(
                insta::internals::AutoName,
                format_debug_result(result),
                $text
            );
        }};
    }

    #[test]
    fn test_parse_pure_text() {
        assert_parse!("Hello name, good to meet you");
    }

    #[test]
    fn test_parse_single_parens() {
        assert_parse!("Hello { name }, good to meet you");
    }

    #[test]
    fn test_parse_identifier() {
        assert_parse!("Hello {{ name }}, good to meet you");
    }

    #[test]
    fn test_parse_gleam_expression() {
        assert_parse!("Hello {{ string.uppercase(name) }}, good to meet you");
    }

    #[test]
    fn test_parse_if_statement() {
        assert_parse!("Hello {% if is_user %}User{% endif %}");
    }

    #[test]
    fn test_parse_empty_if_statement() {
        assert_parse!("Hello {% if is_user %}{% endif %}");
    }

    #[test]
    fn test_parse_if_else_statement() {
        assert_parse!("Hello {% if is_user %}User{% else %}Unknown{% endif %}");
    }

    #[test]
    fn test_parse_if_comparison() {
        assert_parse!("Hello {% if items != [] %}Some items{% endif %}");
    }

    #[test]
    fn test_parse_nested_if_statements() {
        assert_parse!(
            "Hello {% if is_user %}{% if is_admin %}Admin{% else %}User{% endif %}{% endif %}"
        );
    }

    #[test]
    fn test_parse_for_loop_with_single_identifier() {
        assert_parse!("Hello {% for item in list %}{{ item }}{% endfor %}");
    }

    #[test]
    fn test_parse_for_loop() {
        assert_parse!("Hello,{% for item in list %} to {{ item }} and {% endfor %} everyone else");
    }

    #[test]
    fn test_parse_for_as_loop() {
        assert_parse!(
            "Hello,{% for item as Item in list %} to {{ item }} and {% endfor %} everyone else"
        );
    }

    #[test]
    fn test_parse_for_from_expression() {
        assert_parse!("Hello {% for item as Item in list.take(list, 2) %}{{ item }}{% endfor %}");
    }

    #[test]
    fn test_parse_dot_access() {
        assert_parse!("Hello{% if user.is_admin %} Admin{% endif %}");
    }

    #[test]
    fn test_parse_import() {
        assert_parse!("{> import user.{User}\n{{ name }}");
    }

    #[test]
    fn test_parse_with() {
        assert_parse!("{> with user as User\n{{ user }}");
    }

    #[test]
    fn test_parse_builder_block() {
        assert_parse!("Hello {[ name ]}, good to meet you");
    }

    #[test]
    fn test_parse_builder_expression() {
        assert_parse!("Hello {[ string_builder.from_strings([\"Anna\", \" and \", \"Bob\"]) ]}, good to meet you");
    }
}
