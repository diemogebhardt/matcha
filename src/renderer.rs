use crate::parser::Node;
use crate::scanner::Range;

type NodeIter<'a> = std::iter::Peekable<std::slice::Iter<'a, Node>>;

#[derive(Debug)]
pub enum RenderError {
    DuplicateParamName(String, Range),
}

pub fn render(
    iter: &mut NodeIter,
    prog_name: &str,
    from_file_name: &str,
) -> Result<String, RenderError> {
    let (builder_lines, imports, typed_params, includes_for_loop) = render_lines(iter)?;

    let import_lines = imports
        .iter()
        .map(|details| format!("import {}", details))
        .collect::<Vec<_>>()
        .join("\n");

    let params_string = typed_params
        .iter()
        .map(|(param_name, type_name)| format!("{} {}: {}", param_name, param_name, type_name))
        .collect::<Vec<_>>()
        .join(", ");

    let args_string = typed_params
        .iter()
        .map(|(param_name, _)| format!("{}: {}", param_name, param_name))
        .collect::<Vec<_>>()
        .join(", ");

    let list_import = if includes_for_loop {
        "import gleam/list\n"
    } else {
        ""
    };

    let output = format!(
        r#"// DO NOT EDIT: Code generated by {prog_name} from {source_file}

import gleam/string_builder.{{StringBuilder}}
{list_import}
{import_lines}

pub fn render_builder({params_string}) -> StringBuilder {{
    let builder = string_builder.from_string("")
{builder_lines}
    builder
}}

pub fn render({params_string}) -> String {{
    string_builder.to_string(render_builder({args_string}))
}}
"#,
        prog_name = prog_name,
        source_file = from_file_name,
        list_import = list_import,
        import_lines = import_lines,
        params_string = params_string,
        builder_lines = builder_lines,
        args_string = args_string
    );

    Ok(output)
}

type RenderDetails = (String, Vec<String>, Vec<(String, String)>, bool);

fn render_lines(iter: &mut NodeIter) -> Result<RenderDetails, RenderError> {
    let mut builder_lines = String::new();
    let mut imports = vec![];

    // Use a Vec<(String, String)> instead of a HashMap to maintain order which gives the users
    // some control, though parameters are labelled and can be called in any order. Some kind of
    // order is required to keep the tests passing as it seems to be non-determinate in a HashMap
    let mut typed_params = Vec::new();
    let mut includes_for_loop = false;

    loop {
        match iter.peek() {
            Some(Node::Text(text)) => {
                iter.next();
                builder_lines.push_str(&format!(
                    "    let builder = string_builder.append(builder, \"{}\")\n",
                    text.replace("\"", "\\\"")
                ));
            }
            Some(Node::Identifier(name)) => {
                iter.next();
                builder_lines.push_str(&format!(
                    "    let builder = string_builder.append(builder, {})\n",
                    name
                ));
            }
            Some(Node::Builder(name)) => {
                iter.next();
                builder_lines.push_str(&format!(
                    "    let builder = string_builder.append_builder(builder, {})\n",
                    name
                ));
            }
            Some(Node::Import(import_details)) => {
                iter.next();
                imports.push(import_details.clone());
            }
            Some(Node::With((identifier, range), type_)) => {
                iter.next();

                if typed_params.iter().any(|(name, _)| name == identifier) {
                    return Err(RenderError::DuplicateParamName(
                        identifier.clone(),
                        range.clone(),
                    ));
                }

                typed_params.push((identifier.clone(), type_.clone()));
            }
            Some(Node::If(identifier_name, if_nodes, else_nodes)) => {
                iter.next();
                let (if_lines, _, _, if_block_includes_for_loop) =
                    render_lines(&mut if_nodes.iter().peekable())?;
                let (else_lines, _, _, else_block_includes_for_loop) =
                    render_lines(&mut else_nodes.iter().peekable())?;
                builder_lines.push_str(&format!(
                    r#"    let builder = case {} {{
        True -> {{
            {}
            builder
        }}
        False -> {{
            {}
            builder
        }}
}}
"#,
                    identifier_name, if_lines, else_lines
                ));
                includes_for_loop =
                    includes_for_loop || if_block_includes_for_loop || else_block_includes_for_loop;
            }
            Some(Node::For(entry_identifier, entry_type, list_identifier, loop_nodes)) => {
                iter.next();

                let entry_type = entry_type
                    .as_ref()
                    .map(|value| format!(": {}", value))
                    .unwrap_or_else(|| "".to_string());

                let (loop_lines, _, _, _) = render_lines(&mut loop_nodes.iter().peekable())?;
                builder_lines.push_str(&format!(
                    r#"    let builder = list.fold({}, builder, fn(builder, {}{}) {{
        {}
        builder
}})
"#,
                    list_identifier, entry_identifier, entry_type, loop_lines
                ));

                includes_for_loop = true;
            }
            None => break,
        }
    }

    Ok((builder_lines, imports, typed_params, includes_for_loop))
}

#[cfg(test)]
mod test {
    use super::*;

    use crate::parser::{self, ParserError};
    use crate::scanner::{self, ScanError};

    #[derive(Debug)]
    pub enum Error {
        Scan(ScanError),
        Parse(ParserError),
        Render(RenderError),
    }

    fn format_result(result: Result<String, Error>) -> String {
        match result {
            Ok(value) => value,
            Err(err) => format!("{:?}", err),
        }
    }

    const NAME: &str = env!("CARGO_PKG_NAME");

    #[macro_export]
    macro_rules! assert_render {
        ($text:expr $(,)?) => {{
            let _ = env_logger::try_init();
            let result = scanner::scan($text)
                .map_err(|err| Error::Scan(err))
                .and_then(|tokens| {
                    parser::parse(&mut tokens.iter().peekable()).map_err(|err| Error::Parse(err))
                })
                .and_then(|ast| {
                    render(&mut ast.iter().peekable(), NAME, "-test-")
                        .map_err(|err| Error::Render(err))
                });
            insta::assert_snapshot!(insta::internals::AutoName, format_result(result), $text);
        }};
    }

    // Render

    #[test]
    fn test_render_pure_text() {
        assert_render!("Hello name, good to meet you");
    }

    #[test]
    fn test_render_identifier() {
        assert_render!(
            "{> with name as String
Hello {{ name }}, good to meet you"
        );
    }

    #[test]
    fn test_render_two_identifiers() {
        assert_render!(
            "{> with name as String
{> with adjective as String
Hello {{ name }}, {{ adjective }} to meet you"
        );
    }

    #[test]
    fn test_render_gleam_expression() {
        assert_render!(
            "{> import gleam/string
Hello {{ string.uppercase(name) }}, good to meet you"
        );
    }

    #[test]
    fn test_repeated_identifier_usage() {
        assert_render!(
            "{> with name as String
{{ name }} usage, {{ name }} usage"
        );
    }

    #[test]
    fn test_render_if_statement() {
        assert_render!(
            "{> with is_user as Bool
Hello {% if is_user %}User{% endif %}"
        );
    }

    #[test]
    fn test_render_empty_if_statement() {
        assert_render!(
            "{> with is_user as Bool
Hello {% if is_user %}{% endif %}"
        );
    }

    #[test]
    fn test_render_if_else_statement() {
        assert_render!(
            "{> with is_user as Bool
Hello {% if is_user %}User{% else %}Unknown{% endif %}"
        );
    }

    #[test]
    fn test_render_if_comparison() {
        assert_render!("Hello {% if items != [] %}Some items{% endif %}");
    }

    #[test]
    fn test_render_nested_if_statements() {
        assert_render!(
            "{> with is_user as Bool
{> with is_admin as Bool
Hello {% if is_user %}{% if is_admin %}Admin{% else %}User{% endif %}{% endif %}"
        );
    }

    #[test]
    fn test_render_for_loop() {
        assert_render!(
            "{> with list as List(String)
Hello,{% for item in list %} to {{ item }} and {% endfor %} everyone else"
        );
    }

    #[test]
    fn test_render_for_as_loop() {
        assert_render!(
            "{> with list as List(Item)
Hello,{% for item as Item in list %} to {{ item }} and {% endfor %} everyone else"
        );
    }

    #[test]
    fn test_render_for_from_expression() {
        assert_render!("Hello {% for item as Item in list.take(list, 2) %}{{ item }}{% endfor %}");
    }

    #[test]
    fn test_render_dot_access() {
        assert_render!(
            "{> with user as MyUser
Hello{% if user.is_admin %} Admin{% endif %}"
        );
    }

    #[test]
    fn test_render_import() {
        assert_render!("{> import user.{User}\n{> with name as String\n{{ name }}");
    }

    #[test]
    fn test_render_with() {
        assert_render!("{> with user as User\n{{ user }}");
    }

    #[test]
    fn test_render_import_and_with() {
        assert_render!("{> import user.{User}\n{> with user as User\n{{ user }}");
    }

    #[test]
    fn test_render_multiline() {
        assert_render!(
            r#"{> with my_list as List(String)
<ul>
{% for entry in my_list %}
    <li>{{ entry }}</li>
{% endfor %}
</ul>"#
        );
    }

    #[test]
    fn test_render_quotes() {
        assert_render!(
            r#"{> with name as String
<div class="my-class">{{ name }}</div>"#
        );
    }

    #[test]
    fn test_render_builder_block() {
        assert_render!(
            "{> with name as StringBuilder
Hello {[ name ]}, good to meet you"
        );
    }

    #[test]
    fn test_render_builder_expression() {
        assert_render!("Hello {[ string_builder.from_strings([\"Anna\", \" and \", \"Bob\"]) ]}, good to meet you");
    }
}
