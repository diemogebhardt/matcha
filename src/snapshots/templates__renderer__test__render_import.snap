---
source: src/renderer.rs
assertion_line: 291
expression: "{> import user.{User}\n{> with name as String\n{{ name }}"

---
// DO NOT EDIT: Code generated by templates from -test-

import gleam/string_builder.{StringBuilder}
import gleam/list

import user.{User}

pub fn render_builder(name name: String) -> StringBuilder {
    let builder = string_builder.from_string("")
    let builder = string_builder.append(builder, name)

    builder
}

pub fn render(name name: String) -> String {
    string_builder.to_string(render_builder(name: name))
}

